src/mpi/coll/bcast/bcast.c

```diff
@@ -85,7 +85,6 @@
     - name        : MPIR_CVAR_BCAST_INTRA_ALGORITHM
       category    : COLLECTIVE
       type        : enum
-      group       : MPIR_CVAR_GROUP_COLL_ALGO
```

Comment:
Intel requested this feature. They have some internal code that requires different collective algorithms in the same enum group -- probably just to save a level of branching. I have no idea why and how these 4 particular cvars are selected to have the group, though.

Search `enum_groups` in `extractcvars.in`.

Reply:
I'll let the Intel folks comment on this (@wesbland?), but this is an arbitrary collection of CVARs.  Either it should be done consistently or not at all.  I'm not going to leave some arbitrary hack inside the code.

Reply:
cc: @zhenggb72

Reply:
Yes, we requested this in the hackathon last year. 
The idea is that all enum in the same enum group will have a unique integer number for each algorithm in this case.

Reply:
Why do you need this?  The CVAR is already algorithm specific.  Why do you need it to be unique across the entire group?

Reply:
otherwise every collectives starts enum from 0, which makes it harder to detect user error when they specify a wrong algorithm.

Reply:
What kind of user errors are you trying to catch?  Can you be more precise?

Reply:
export MPIR_CVAR_BCAST_INTRA_ALGORITHM=MPIR_CVAR_ALLREDUCE_POSIX_INTRA_ALGORITHM_json

for example, a user can give an allreduce algorithm, and our Bcast_impl big switch statement can not catch that.

Reply:
sorry, wrong example. let me think about it.

Reply:
That's not how it's specified.  It's specified as `export MPIR_CVAR_BCAST_INTRA_ALGORITHM=json`.  So your use case is incorrect.

Reply:
ok, it was an idea we didn't pursue further. We tried to think about a way when the CVAR algorithm is specified at OFI, but the algorithm is implemented in MPIR layer. In current implementation, the OFI CVAR alos lists all algorithm in MPIR. Ideally OFI only needs to enumerate device level algorithm. Having this enum group, potentially allows enums in MPIR and OFI to be combined into one.

Reply:
No, it doesn't allow them to be combined.  OFI is not the only netmod.  And ch4 is not the only device.   When you folks come up with a concrete proposal, we can reconsider this.  Till then, I can't have this code in `mpich/master`.

Reply:
Combining enums to one struct based on group does not have side effects (it only changes the enum integer value), it potentially can be used by OFI to pass the CVAR down to MPIR to re-use the MPIR level CVAR-based selection, and so to avoid the situation that each upper layer has to repeat the CVAR-based selection for MPIR. This is a problem to consider if you are interested. But right now we are fine without the group since we repeat MPIR selection in OFI.

Reply:
`Combining enums to one struct based on group does not have side effects`

The side effect is more code maintenance.  I had to spend more time understanding my code to add the enumlists.  So please don't ever say that adding unused code is OK.  It's a software maintenance nightmare.

`it potentially can be used by OFI to pass the CVAR down to MPIR to re-use the MPIR level CVAR-based selection`

OFI shouldn't have to pick any specific MPIR-level algorithm if the MPIR-level algorithms are tuned.  But even if it did, all the MPIR-level algorithms are already visible to OFI.  Why do you want the user to provide an OFI-level CVAR to specify an MPIR-level algorithm?

Reply:
It was a simple idea when we discussed the issue of selection during Hackathon, and it happened Hui implemented that feature (group) in just like a minute, so he checked it in. Like I said, if we want to handle the OFI selection differently, we will use that feature. Right now, it is not our priority. 

For the second question, there are several methods we thought about to avoid OFI specifying MPIR algorithms, but there are pros and cons of each solution. For example, for BCAST, there is MPIR_CVAR_BCAST_OFI_INTRA_ALGORITHM for OFI and MPIR_CVAR_BCAST_INTRA_ALGORITHM for MPIR. The question is if you want to allow a user to say MPIR_CVAR_BCAST_OFI_INTRA_ALGORITHM=binomial, which is a MPIR-level algorithm. Currently, to make it easy for mpitune, we choose to allow that.  Another possible way is to say:  MPIR_CVAR_BCAST_OFI_INTRA_ALGORITHM=none  (or some string which means bypassing the ofi selection) and together with MPIR_CVAR_BCAST_INTRA_ALGORITHM=binomial.  But this requires users to know which algorithm is at which level which is tedious. Now if choose to allow MPIR_CVAR_BCAST_OFI_INTRA_ALGORITHM=binomial, then all enum of MPIR-level algorithm needs to be in OFI enum block, which is a duplication. That Hui's patch on supporting 'group' was part of the idea we tried to reduce this duplication, but we haven't got a chance to implement our part yet.

We can discuss this more when you are here, but like I said. I am ok with this change for now.

Reply:
It took me an hour to work through the enum groups while I was adding enumlists, which is time I had to waste because of adding a feature that's not actually used just.  And now we are wasting more time arguing about it.  😄 

Why do you want the OFI-level CVAR to every specify an MPIR-level algorithm?  Use the MPIR-level CVAR for it.  I don't understand MPITUNE well enough to comment about it, but I feel like once the MPIR-level algorithms are tuned, the OFI-level algorithms should not be picking specific MPIR-level algorithms.

Reply:
if to make enumlists useful (which both Surabhi and I have not found a use case for us yet), you would want OFI-level CVAR to specify an MPI-level one, otherwise, how you can specify a fallback to MPIR-level algorithm in that OFI CVAR using enum list?

Reply:
I said you don't need specific MPIR-level algorithms.  I didn't say you shouldn't call MPIR-level collectives at all.  In other words, you can still call `MPIR_Bcast` or `MPIR_Bcast_impl`, but I don't see why you need to call `MPIR_Bcast_binomial`.